### ISSUE TITLE
The XS/C++ code failed to compile

### ISSUE THREAD

**Issue by megalomaniac** [alexrj/Slic3r#1690]
I'm using ubuntu 13.10 64bit was a fresh install.

followed the instructions on https://github.com/alexrj/Slic3r/wiki/Running-Slic3r-from-git-on-GNU-Linux
Didn't work I thought it was a perl problem so I then installed perlbrew perl-5.16.0

After installing perlberw lots of things failed after trying sudo perl Build.PL --gui

Then I Tried perl Build.PL  everything tested ok except 'The XS/C++ code failed to compile'   
Then I tried with the following --force command and got this:

:~/Slic3r$ sudo perl Build.PL --force
App::cpanminus is up to date. (1.7001)
Class::XSAccessor is up to date. (1.19)
Encode::Locale is up to date. (1.03)
ExtUtils::MakeMaker is up to date. (6.86)
ExtUtils::ParseXS is up to date. (3.22)
skipping R/RJ/RJBS/perl-5.18.2.tar.bz2
File::Spec is up to date. (3.40)
Getopt::Long is up to date. (2.42)
IO::Scalar is up to date. (2.110)
Math::Geometry::Voronoi is up to date. (1.3)
Math::PlanePath is up to date. (113)
Module::Build::WithXSpp is up to date. (0.14)
Moo is up to date. (1.004002)
Scalar::Util is up to date. (1.35)
Storable is up to date. (2.45)
Test::Harness is up to date. (3.30)
Test::More is up to date. (1.001002)
Time::HiRes is up to date. (1.9726)
XML::SAX::ExpatXS is up to date. (1.33)
Cleaning up build files
Cleaning up configuration files
--> Working on ./xs
Configuring /home/mega/Slic3r/xs ... OK
Building and testing Slic3r-XS-0.01 ... FAIL
! Installing ./xs failed. See /home/mega/.cpanm/work/1389327043.19782/build.log for details. Retry with --force to force install it.
The XS/C++ code failed to compile, aborting

Then when I try this
./slic3r.pl 
Can't locate Slic3r/XS.pm in @INC (@INC contains: /home/mega/Slic3r/lib /etc/perl /usr/local/lib/perl/5.14.2 /usr/local/share/perl/5.14.2 /usr/lib/perl5 /usr/share/perl5 /usr/lib/perl/5.14 /usr/share/perl/5.14 /usr/local/lib/site_perl .) at /home/mega/Slic3r/lib/Slic3r.pm line 38.
BEGIN failed--compilation aborted at /home/mega/Slic3r/lib/Slic3r.pm line 38.
Compilation failed in require at ./slic3r.pl line 14.
BEGIN failed--compilation aborted at ./slic3r.pl line 14.

Don't know what to try now.

**Comment by lerouxb** [comment]
Trying to build on OSX Mavericks, getting:

````
cpanm (App::cpanminus) 1.7027 on perl 5.016002 built for darwin-thread-multi-2level
Work directory is /Users/leroux/.cpanm/work/1427484056.10688
You have make /usr/bin/make
You have LWP 6.04
You have /usr/bin/tar: bsdtar 2.8.3 - libarchive 2.8.3
You have /usr/bin/unzip
--> Working on ./xs
Entering /Users/leroux/src/Slic3r/xs
Checking if you have Module::Build 0.36 ... Yes (0.4211)
Configuring /Users/leroux/src/Slic3r/xs
Running Build.PL
-> N/A
-> FAIL Configure failed for ./xs. See /Users/leroux/.cpanm/work/1427484056.10688/build.log for details.
````

**Comment by lordofhyphens** [comment]
It'd help if you included the  contents of /Users/leroux/.cpanm/work/
1427484056.10688/build.log

"Nothing unreal exists." - Kiri-kin-tha's First Law of Metaphysics.

**Comment by lerouxb** [comment]
That IS the contents of that file. It just happens to reference itself at the end.

**Comment by alexrj** [comment]
@lerouxb, try manually:

```
cd xs
sudo perl Build.PL
sudo ./Build install
```

If it fails, post output here...

**Comment by gralco** [comment]
I'm on Fedora 22, when I try `perl Build.PL --sudo` just inside `./Slic3r/` I get the same **`The XS/C++ code failed to compile, aborting`**

When I do what alexrj said I get no output whatsoever.

**Comment by Tuxomx** [comment]
Hi all. I had the same problem, then used strace to check perl compiling xs. Perl was searching for some files, so I installed them and no luck. On a second view of the strace I found perl was using some files from /usr/local/lib64 remaining from fedora 20 (upgraded to 22 today), erasing all of the /usr/local/lib65/perl5/ files allowed me to run "stock" slic3r (the one installed by default) and compile current version (got with git clone https://github.com/alexrj/Slic3r.git as stated on README.md).

SALUDOS
ALVARO

**Comment by lordofhyphens** [comment]
Ah, CPAN hell. Yeah, the basic compile instructions tend to make your perl
install fragile w/r/t package managers in distros. You really should be
using your package manager  (that is, building an rpm from slic3r source
and installing that). A shortcut if someone has already packaged an older
version of slic3r is to grab the source for that package and unpack the new
slic3r source into it (on top of the old package source) and then run the
package build scripts.

**Comment by alexrj** [comment]
@gralco, no output while typing the three commands above?  I can't really believe that...

**Comment by gralco** [comment]
@alexrj 
No output.

Also, when I perform `sudo perl Build.PL` there is no generated "Build" file.

Take a look:

```
[ekuzmenko@localhost Slic3r]$ cd xs
[ekuzmenko@localhost xs]$ sudo perl Build.PL 
[ekuzmenko@localhost xs]$ sudo ./Build install
sudo: ./Build: command not found
[ekuzmenko@localhost xs]$ sudo ./Build.PL install
sudo: ./Build.PL: command not found
[ekuzmenko@localhost xs]$ sudo chmod 777 ./Build.PL 
[ekuzmenko@localhost xs]$ sudo ./Build.PL install
[ekuzmenko@localhost xs]$ ls
Build.PL  lib  MANIFEST  MANIFEST.SKIP  src  t  xsp
```

**Comment by HustLion** [comment]
@alexrj  I guess the user can't build slic3r with perl 5.16 or higher, recommend use Perl-5.14.4


Those people above used high version perl, such as the first one @megalomaniac with perl-5.16.0, which definitely can't work. And @bmcage  used 5.18 I guess ( in this comment there is a line `cc -I/usr/lib/perl/5.18/CORE -DXS_VERSION="0.01" -DVERSION="0.01" -fPIC -xc++ `

Actually the newer version of Linux the new developers are using, the higher the possiblity they will encounter this problem and get upset. Because newer linux distributions ship perl 5.16 or even 5.18.

```` bash
curl -L http://xrl.us/perlbrewinstall | bash
source ~/perl5/perlbrew/etc/bashrc
perlbrew install 5.14.4
perlbrew switch perl-5.14.4
perlbrew install-cpanm
git clone https://github.com/alexrj/Slic3r.git
cd Slic3r
CPANM=/home/hustlion/perl5/perlbrew/bin/cpanm perl Build.PL
CPANM=/home/hustlion/perl5/perlbrew/bin/cpanm perl Build.PL --gui
./slic3r.pl
````

I have updated the wiki [page](https://github.com/alexrj/Slic3r/wiki/Running-Slic3r-from-git-on-GNU-Linux) accroding to this. The CPANM path is changed (replace /home/hustlion/ by ~/).

Hope this can be helpful :smile:.

**Comment by bmcage** [comment]
Asking people to use 5.14 via perlbrew with no automatic updates??? That is not a solution.

Sofware should be able to use Ubuntu 14.04 default packages. It has been a long time that I used Slic3r, but I got it running on Ubuntu 14.04 with perl 5.18 from one of the devel branches.

**Comment by lerouxb** [comment]
I'm on Perl v5.12.5 on OSX Mavericks and having this issue, so I'm not convinced that's the problem.

**Comment by HustLion** [comment]
@bmcage Oh, I'm just offering a way to run the one on the master branch. Yes I'm thinking Slic3r should be able to run with the default one not just 5.14. Acutally two months ago I cloned Slic3r and can't compile it and finally found out that it's the version that I'm using that causes the problem. Actually I found it out from the source code, I can't remember where it is. But the prompt is `Running Slic3r under Perl >= 5.16 is not supported nor recommended`. If you are still not convinced then I may find the exact line for you :stuck_out_tongue_closed_eyes: .

@lerouxb So,,, maybe your version is too old?

**Comment by HustLion** [comment]
@lerouxb By the way, I only tested the method with Ubuntu 14.04. So I don't know if it will work for you on OSX Mavericks. And your case meets the requirement that Perl <5.16, so,,, maybe you can try again and give some feedback about it?

**Comment by lerouxb** [comment]
I have pulled the latest changes and ran the install again as specified in the wiki:
https://github.com/alexrj/Slic3r/wiki/Running-Slic3r-from-git-on-OS-X

(Note that the wiki tells you that 5.16 won't work, so it tells you to set the default to 5.12 which I did, but then the error mentions 5.16 again, so I don't know what's going on there or how to force it to use the right version of perl.)

Here is the result of running it:
https://gist.github.com/lerouxb/bbe62b83e09ff5175779

The attached build.log is the contents of the build.log file.

**Comment by HustLion** [comment]
@lerouxb That wiki page you mentioned is not updated by me. And this is the one: https://github.com/alexrj/Slic3r/wiki/Running-Slic3r-from-git-on-GNU-Linux

Anyway, from your description, I can only offer a way to easily choose the version of perl that you will to run.

The solution is to use [perlbrew](http://perlbrew.pl/). With perlbrew, you can use `perlbrew available` to check the version of perl that is available for installation. And use `perlbrew install 5.14.4` to install the version you choose (I use perl-5.14.4 as an example). This may take a while. After the installation is complete, you can use `perlbrew switch perl-5.14.4` to change the version of perl to 5.14.4. Now use `perl -v` to check the version of your perl, it should be 5.14.4. If you want to switch back, you can just turn off perlbrew by `perlbrew off`, at this time if you run `perl -v`, it should be the version of perl that ships with your system.

To sum it up, use perlbrew to switch between different versions of perl. And try to compile again with perl-5.14.4. You can find the installation instructions on [perlbrew homepage](http://perlbrew.pl/). I suggest you read how I deal with it in Ubuntu https://github.com/alexrj/Slic3r/wiki/Running-Slic3r-from-git-on-GNU-Linux. The solution may be similar.

And, sorry that I can't open the gist link that you provided due to our Great Fire Wall (Chinese,, hmm, :cry: .

**Comment by lordofhyphens** [comment]
In my experience, perl 5.20.1 compiles Slic3r 1.2.x fine on Linux (and
Windows via StrawberryPerl). That's what I use for the basis of my Debian
bleeding-edge package (source is available on my fork, binary package at
https://drive.google.com/drive/u/1/folders/0B3aPl7CmzLRvfjBaQmNXNUFlbXlHNHhlLWQxMjQ3X0NjX2p3V3k1LU9EWktkTXNDcWgxeG8).
I am more likely to believe that issues compiling are related to other
issues in the system (libraries, broken CPAN, etc).

5.14 was released 2 years ago *and* is EOL'd (and isn't even shipped by
Debian Stable, one of the crustiest distros in terms of package age).



"Nothing unreal exists." - Kiri-kin-tha's First Law of Metaphysics.

**Comment by alexrj** [comment]
Actually, Slic3r works fine with all Perl versions except for 5.16.

**Comment by alexrj** [comment]
Unfortunately, I don't have access to a Maverick machine so I can't test myself. Maybe a user-land perl  installed with perlbrew helps.

**Comment by demolition602099** [comment]
![image](https://cloud.githubusercontent.com/assets/19219093/15134382/76792356-1620-11e6-99fa-ec9896046a92.png)

Hi!,

I am having  trouble compiling the slic3r. I have used Strawberry Pearl and Citrus Perl.


Thank you in advance

**Comment by lordofhyphens** [comment]
Yeah, odds are your exact version of Perl is misconfigured or not
compatible. Gg.

**Comment by demolition602099** [comment]
My strawberry Pearl 5.14.
How can i change my configuration?
Thank You for the reply

**Comment by a4jp-com** [comment]
> Microsoft Windows [Version 10.0.10586]
(c) 2015 Microsoft Corporation. All rights reserved.

C:\Users\design>perl -v

This is perl 5, version 22, subversion 0 (v5.22.0) built for MSWin32-x86-multi-thread-64int

Copyright 1987-2015, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.


> 

I use the 32 bit version of Perl v5.22 and everything complies well.

**Comment by demolition602099** [comment]
What kind of installer did you use?
Strawberry? or something?
Thanks

**Comment by demolition602099** [comment]
I have made it working.
Mine was Windows 10 64bit.
Strawberry 5.22.1.3 64Bit Version
MingW 64 for perl ( Downloaded at source forge)

**Comment by lordofhyphens** [comment]
Closing the issue then.

**Comment by dergrunepunkt** [comment]
Fixed with:

```
perl -MCPAN -e shell
CPAN> o conf make /usr/bin/make
```
